File: lib/relation.rb
require 'ast/builder'
require 'definition'
require 'rule'

class Relation
  def initialize(name, definition)
    self.name, self.definition = name, definition
  end

  def self.define(name:, rules:)
    new(name, Definition.new(rules.map(&Rule.method(:define))))
  end

  NoRuleApplies = Class.new(StandardError)
  Nondeterministic = Class.new(StandardError)

  def once(*inputs)
    builder = AST::Builder.new
    output = builder.build_variable('output')
    keywords = Array(name).map(&builder.method(:build_keyword))
    formula = builder.build_sequence(inputs.zip(keywords).flatten(1) + [output])
    states = definition.derive(formula)

    raise NoRuleApplies if states.empty?
    raise Nondeterministic, states.map { |s| s.value_of(result) } if states.length > 1

    states.first.value_of(output)
  end

  def many(input)
    begin
      many(once(input))
    rescue NoRuleApplies
      input
    end
  end

  private

  attr_accessor :name, :definition
end

File: src/relation.js
import Builder from './builder';
import Definition from './definition';
import Rule from './rule';

class NoRuleApplies extends Error {
  constructor(message) {
    super(message);
    this.name = 'NoRuleApplies';
  }
}

class Nondeterministic extends Error {
  constructor(message) {
    super(message);
    this.name = 'Nondeterministic';
  }
}

export default class Relation {
  constructor(name, definition) {
    this.name = name;
    this.definition = definition;
  }

  static define({ name, rules }) {
    return new Relation(name, new Definition(rules.map(Rule.define)));
  }

  once(input) {
    const builder = new Builder();
    const output = builder.buildVariable('output');
    const formula = builder.buildSequence([
      input,
      builder.buildKeyword(this.name),
      output
    ]);
    const states = this.definition.derive(formula);

    if (states.isEmpty()) {
      throw new NoRuleApplies();
    }

    if (states.size > 1) {
      throw new Nondeterministic();
    }

    return states.first().valueOf(output);
  }

  many(input) {
    try {
      return this.many(this.once(input));
    } catch (e) {
      if (e instanceof NoRuleApplies) {
        return input;
      } else {
        throw e;
      }
    }
  }
}

export { NoRuleApplies, Nondeterministic, Relation };
